### PR TITLE
add TensorBoard logging with loss and wps

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #57

Each rank build its own TensorBoard writer. The global loss is communicated among all ranks before logging.

To visualize using SSH tunneling:
`ssh -L 6006:127.0.0.1:6006 your_user_name@my_server_ip`
in torchtrain repo
`tensorboard --logdir=./torchtrain/outputs/tb`
then on web browser go to http://localhost:6006/

Screenshot 2024-02-12 at 6 39 28 PM

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -184,6 +194,32 @@ def main(args):
...
+                avg_loss, max_loss = np.mean(losses_since_last_log), np.max(
+                    losses_since_last_log
+                )
+                global_avg_loss, global_max_loss = dist_mean(
+                    avg_loss, world_mesh
+                ), dist_max(max_loss, world_mesh)
+
+                time_delta = timer() - time_last_log
+                wps = nwords_since_last_log / (
+                    time_delta * parallel_dims.sp * parallel_dims.pp
```
Comment (wanchaol) [inline train.py]: a neater way is to define a `model_parallel_size` in the `parallel_dims` class that return this number directly (i.e. a cached property)
